File: android/app/src/main/java/io/karpilabs/simplemp3/data/jellyfin/JellyfinClient.kt
package io.karpilabs.simplemp3.data.jellyfin

import com.squareup.moshi.Moshi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File
import java.io.IOException
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class JellyfinClient @Inject constructor(
    private val okHttp: OkHttpClient,
    private val moshi: Moshi
) {
    private val authAdapter by lazy { moshi.adapter(AuthenticationResult::class.java) }
    private val queryAdapter by lazy { moshi.adapter(QueryResult::class.java) }
    private val itemAdapter by lazy { moshi.adapter(JellyfinItem::class.java) }
    private val authRequestAdapter by lazy { moshi.adapter(AuthenticateByNameRequest::class.java) }

    fun normalizeServerUrl(raw: String): String {
        var url = raw.trim().trimEnd('/')
        if (!url.startsWith("http://") && !url.startsWith("https://")) {
            url = "http://$url"
        }
        return url
    }

    private fun authHeader(deviceId: String, token: String? = null): String {
        val base = buildString {
            append("""MediaBrowser Client="SimpleMP3", Device="Android", DeviceId="$deviceId", Version="1.0.0"""")
            if (!token.isNullOrBlank()) {
                append(""", Token="$token"""")
            }
        }
        return base
    }

    suspend fun authenticate(
        serverUrl: String,
        username: String,
        password: String,
        deviceId: String
    ): Result<JellyfinSession> = withContext(Dispatchers.IO) {
        runCatching {
            val base = normalizeServerUrl(serverUrl)
            val bodyJson = authRequestAdapter.toJson(
                AuthenticateByNameRequest(username = username.trim(), password = password)
            )
            val request = Request.Builder()
                .url("$base/Users/AuthenticateByName")
                .post(bodyJson.toRequestBody("application/json".toMediaType()))
                .header("X-Emby-Authorization", authHeader(deviceId))
                .header("Content-Type", "application/json")
                .header("Accept", "application/json")
                .build()

            okHttp.newCall(request).execute().use { response ->
                val body = response.body?.string().orEmpty()
                if (!response.isSuccessful) {
                    throw IOException("Login failed (${response.code}): ${body.take(200)}")
                }
                val parsed = authAdapter.fromJson(body)
                    ?: throw IOException("Invalid login response")
                val token = parsed.accessToken
                    ?: throw IOException("No access token returned")
                val user = parsed.user
                    ?: throw IOException("No user in response")
                JellyfinSession(
                    serverUrl = base,
                    accessToken = token,
                    userId = user.id,
                    userName = user.name.orEmpty().ifBlank { username },
                    serverId = parsed.serverId,
                    deviceId = deviceId
                )
            }
        }
    }

    suspend fun getAudioItems(
        session: JellyfinSession,
        startIndex: Int = 0,
        limit: Int = 200,
        parentId: String? = null,
        searchTerm: String? = null
    ): Result<QueryResult> = withContext(Dispatchers.IO) {
        runCatching {
            val params = mutableListOf(
                "IncludeItemTypes=Audio",
                "Recursive=true",
                "Fields=BasicSyncInfo,PrimaryImageAspectRatio,Path,MediaSources",
                "SortBy=Album,IndexNumber,SortName",
                "SortOrder=Ascending",
                "StartIndex=$startIndex",
                "Limit=$limit",
                "EnableImageTypes=Primary"
            )
            if (!parentId.isNullOrBlank()) {
                params += "ParentId=${enc(parentId)}"
            }
            if (!searchTerm.isNullOrBlank()) {
                params += "SearchTerm=${enc(searchTerm)}"
            }
            val url = "${session.serverUrl}/Users/${session.userId}/Items?${params.joinToString("&")}"
            getJson(session, url, queryAdapter) ?: QueryResult()
        }
    }

    suspend fun getAlbums(
        session: JellyfinSession,
        startIndex: Int = 0,
        limit: Int = 100
    ): Result<QueryResult> = withContext(Dispatchers.IO) {
        runCatching {
            val url = buildString {
                append(session.serverUrl)
                append("/Users/${session.userId}/Items?")
                append("IncludeItemTypes=MusicAlbum&Recursive=true")
                append("&SortBy=SortName&SortOrder=Ascending")
                append("&StartIndex=$startIndex&Limit=$limit")
                append("&EnableImageTypes=Primary")
                append("&Fields=ChildCount,PrimaryImageAspectRatio")
            }
            getJson(session, url, queryAdapter) ?: QueryResult()
        }
    }

    suspend fun getAlbumTracks(
        session: JellyfinSession,
        albumId: String
    ): Result<List<JellyfinItem>> = withContext(Dispatchers.IO) {
        runCatching {
            val url = buildString {
                append(session.serverUrl)
                append("/Users/${session.userId}/Items?")
                append("ParentId=${enc(albumId)}")
                append("&IncludeItemTypes=Audio")
                append("&Recursive=true")
                append("&SortBy=IndexNumber,SortName")
                append("&SortOrder=Ascending")
                append("&Fields=BasicSyncInfo,PrimaryImageAspectRatio")
                append("&Limit=500")
            }
            getJson(session, url, queryAdapter)?.items.orEmpty()
        }
    }

    suspend fun getPlaylists(
        session: JellyfinSession,
        startIndex: Int = 0,
        limit: Int = 100
    ): Result<QueryResult> = withContext(Dispatchers.IO) {
        runCatching {
            val url = buildString {
                append(session.serverUrl)
                append("/Users/${session.userId}/Items?")
                append("IncludeItemTypes=Playlist&Recursive=true")
                append("&SortBy=SortName&SortOrder=Ascending")
                append("&StartIndex=$startIndex&Limit=$limit")
                append("&EnableImageTypes=Primary")
                append("&Fields=ChildCount,PrimaryImageAspectRatio")
            }
            getJson(session, url, queryAdapter) ?: QueryResult()
        }
    }

    suspend fun getPlaylistTracks(
        session: JellyfinSession,
        playlistId: String
    ): Result<List<JellyfinItem>> = withContext(Dispatchers.IO) {
        runCatching {
            val url = buildString {
                append(session.serverUrl)
                append("/Playlists/${enc(playlistId)}/Items?")
                append("UserId=${enc(session.userId)}")
                append("&Fields=BasicSyncInfo,PrimaryImageAspectRatio,MediaSources")
                append("&EnableImageTypes=Primary")
                append("&Limit=500")
            }
            getJson(session, url, queryAdapter)?.items.orEmpty()
        }
    }

    suspend fun getItem(
        session: JellyfinSession,
        itemId: String
    ): Result<JellyfinItem> = withContext(Dispatchers.IO) {
        runCatching {
            val url = "${session.serverUrl}/Users/${session.userId}/Items/${enc(itemId)}" +
                "?Fields=BasicSyncInfo,PrimaryImageAspectRatio,MediaSources,Path"
            getJson(session, url, itemAdapter)
                ?: throw IOException("Item not found")
        }
    }

    fun imageUrl(session: JellyfinSession, item: JellyfinItem, maxWidth: Int = 400): String? {
        val imageItemId = when {
            item.imageTags?.containsKey("Primary") == true -> item.id
            !item.albumId.isNullOrBlank() && !item.albumPrimaryImageTag.isNullOrBlank() -> item.albumId
            else -> return null
        }
        return "${session.serverUrl}/Items/$imageItemId/Images/Primary?maxWidth=$maxWidth&quality=85&api_key=${session.accessToken}"
    }

    fun streamUrl(session: JellyfinSession, itemId: String): String {
        return "${session.serverUrl}/Audio/$itemId/stream?static=true&api_key=${session.accessToken}"
    }

    suspend fun downloadToFile(
        session: JellyfinSession,
        itemId: String,
        dest: File,
        onProgress: (bytesRead: Long, total: Long?) -> Unit = { _, _ -> }
    ): Result<File> = withContext(Dispatchers.IO) {
        runCatching {
            dest.parentFile?.mkdirs()
            val tmp = File(dest.parentFile, "${dest.name}.part")
            if (tmp.exists()) tmp.delete()

            val request = Request.Builder()
                .url(streamUrl(session, itemId))
                .header("X-Emby-Authorization", authHeader(session.deviceId, session.accessToken))
                .header("X-Emby-Token", session.accessToken)
                .get()
                .build()

            okHttp.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    throw IOException("Download failed (${response.code})")
                }
                val body = response.body ?: throw IOException("Empty body")
                val total = body.contentLength().takeIf { it > 0 }
                body.byteStream().use { input ->
                    tmp.outputStream().use { output ->
                        val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
                        var readTotal = 0L
                        while (true) {
                            val read = input.read(buffer)
                            if (read < 0) break
                            output.write(buffer, 0, read)
                            readTotal += read
                            onProgress(readTotal, total)
                        }
                    }
                }
            }
            if (dest.exists()) dest.delete()
            if (!tmp.renameTo(dest)) {
                tmp.copyTo(dest, overwrite = true)
                tmp.delete()
            }
            dest
        }
    }

    suspend fun downloadImage(
        session: JellyfinSession,
        item: JellyfinItem,
        dest: File
    ): Result<File?> = withContext(Dispatchers.IO) {
        runCatching {
            val url = imageUrl(session, item, maxWidth = 600) ?: return@runCatching null
            dest.parentFile?.mkdirs()
            val request = Request.Builder()
                .url(url)
                .header("X-Emby-Authorization", authHeader(session.deviceId, session.accessToken))
                .get()
                .build()
            okHttp.newCall(request).execute().use { response ->
                if (!response.isSuccessful) return@runCatching null
                val body = response.body ?: return@runCatching null
                dest.outputStream().use { out -> body.byteStream().copyTo(out) }
            }
            dest
        }
    }

    private fun <T> getJson(session: JellyfinSession, url: String, adapter: com.squareup.moshi.JsonAdapter<T>): T? {
        val request = Request.Builder()
            .url(url)
            .header("X-Emby-Authorization", authHeader(session.deviceId, session.accessToken))
            .header("X-Emby-Token", session.accessToken)
            .header("Accept", "application/json")
            .get()
            .build()
        okHttp.newCall(request).execute().use { response ->
            val body = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw IOException("Request failed (${response.code}): ${body.take(200)}")
            }
            return adapter.fromJson(body)
        }
    }

    private fun enc(value: String): String =
        URLEncoder.encode(value, StandardCharsets.UTF_8.name())
}

File: ios/Simple MP3/Jellyfin/JellyfinClient.swift
//
//  JellyfinClient.swift
//  Simple MP3
//

import Foundation

enum JellyfinError: LocalizedError {
    case invalidURL
    case loginFailed(String)
    case notAuthenticated
    case http(Int, String)
    case decode
    case downloadFailed

    var errorDescription: String? {
        switch self {
        case .invalidURL: return "Invalid server URL"
        case .loginFailed(let m): return m
        case .notAuthenticated: return "Not signed in to Jellyfin"
        case .http(let c, let m): return "HTTP \(c): \(m)"
        case .decode: return "Could not parse server response"
        case .downloadFailed: return "Download failed"
        }
    }
}

actor JellyfinClient {
    private let session: URLSession

    init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 600
        session = URLSession(configuration: config)
    }

    func normalizeServerUrl(_ raw: String) -> String {
        var url = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        while url.hasSuffix("/") { url.removeLast() }
        if !url.lowercased().hasPrefix("http://") && !url.lowercased().hasPrefix("https://") {
            url = "http://\(url)"
        }
        return url
    }

    private func authHeader(deviceId: String, token: String? = nil) -> String {
        var s = #"MediaBrowser Client="SimpleMP3", Device="iOS", DeviceId="\#(deviceId)", Version="1.0.0""#
        if let token, !token.isEmpty {
            s += #", Token="\#(token)""#
        }
        return s
    }

    func authenticate(
        serverUrl: String,
        username: String,
        password: String,
        deviceId: String
    ) async throws -> JellyfinSession {
        let base = normalizeServerUrl(serverUrl)
        guard let url = URL(string: "\(base)/Users/AuthenticateByName") else {
            throw JellyfinError.invalidURL
        }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.setValue(authHeader(deviceId: deviceId), forHTTPHeaderField: "X-Emby-Authorization")
        let body = AuthenticateByNameRequest(Username: username.trimmingCharacters(in: .whitespaces), Pw: password)
        req.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse else { throw JellyfinError.loginFailed("No response") }
        if !(200..<300).contains(http.statusCode) {
            let msg = String(data: data, encoding: .utf8)?.prefix(200) ?? ""
            throw JellyfinError.loginFailed("Login failed (\(http.statusCode)): \(msg)")
        }
        let parsed = try JSONDecoder().decode(AuthenticationResult.self, from: data)
        guard let token = parsed.AccessToken, let user = parsed.User else {
            throw JellyfinError.loginFailed("Invalid login response")
        }
        return JellyfinSession(
            serverUrl: base,
            accessToken: token,
            userId: user.Id,
            userName: user.Name?.isEmpty == false ? user.Name! : username,
            serverId: parsed.ServerId,
            deviceId: deviceId
        )
    }

    func getAudioItems(
        session jf: JellyfinSession,
        startIndex: Int = 0,
        limit: Int = 200,
        parentId: String? = nil,
        searchTerm: String? = nil
    ) async throws -> QueryResult {
        var params = [
            "IncludeItemTypes=Audio",
            "Recursive=true",
            "Fields=BasicSyncInfo,PrimaryImageAspectRatio,Path,MediaSources",
            "SortBy=Album,IndexNumber,SortName",
            "SortOrder=Ascending",
            "StartIndex=\(startIndex)",
            "Limit=\(limit)",
            "EnableImageTypes=Primary"
        ]
        if let parentId, !parentId.isEmpty {
            params.append("ParentId=\(parentId.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? parentId)")
        }
        if let searchTerm, !searchTerm.isEmpty {
            params.append("SearchTerm=\(searchTerm.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? searchTerm)")
        }
        let urlStr = "\(jf.serverUrl)/Users/\(jf.userId)/Items?\(params.joined(separator: "&"))"
        return try await getJSON(session: jf, urlString: urlStr)
    }

    func getAlbums(session jf: JellyfinSession, startIndex: Int = 0, limit: Int = 100) async throws -> QueryResult {
        let urlStr = """
        \(jf.serverUrl)/Users/\(jf.userId)/Items?\
        IncludeItemTypes=MusicAlbum&Recursive=true\
        &SortBy=SortName&SortOrder=Ascending\
        &StartIndex=\(startIndex)&Limit=\(limit)\
        &EnableImageTypes=Primary\
        &Fields=ChildCount,PrimaryImageAspectRatio
        """
        return try await getJSON(session: jf, urlString: urlStr)
    }

    func getAlbumTracks(session jf: JellyfinSession, albumId: String) async throws -> [JellyfinItem] {
        let enc = albumId.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? albumId
        let urlStr = """
        \(jf.serverUrl)/Users/\(jf.userId)/Items?\
        ParentId=\(enc)&IncludeItemTypes=Audio&Recursive=true\
        &SortBy=IndexNumber,SortName&SortOrder=Ascending\
        &Fields=BasicSyncInfo,PrimaryImageAspectRatio&Limit=500
        """
        let result: QueryResult = try await getJSON(session: jf, urlString: urlStr)
        return result.Items ?? []
    }

    func getPlaylists(session jf: JellyfinSession, startIndex: Int = 0, limit: Int = 100) async throws -> QueryResult {
        let urlStr = """
        \(jf.serverUrl)/Users/\(jf.userId)/Items?\
        IncludeItemTypes=Playlist&Recursive=true\
        &SortBy=SortName&SortOrder=Ascending\
        &StartIndex=\(startIndex)&Limit=\(limit)\
        &EnableImageTypes=Primary&Fields=ChildCount,PrimaryImageAspectRatio
        """
        return try await getJSON(session: jf, urlString: urlStr)
    }

    func streamURL(session jf: JellyfinSession, itemId: String) -> URL? {
        let enc = itemId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? itemId
        return URL(string: "\(jf.serverUrl)/Audio/\(enc)/stream?static=true&api_key=\(jf.accessToken)")
    }

    func downloadURL(session jf: JellyfinSession, itemId: String) -> URL? {
        streamURL(session: jf, itemId: itemId)
    }

    func imageURL(session jf: JellyfinSession, itemId: String, tag: String?, maxWidth: Int = 300) -> URL? {
        guard let tag, !tag.isEmpty else { return nil }
        let enc = itemId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? itemId
        return URL(string: "\(jf.serverUrl)/Items/\(enc)/Images/Primary?tag=\(tag)&maxWidth=\(maxWidth)&api_key=\(jf.accessToken)")
    }

    func download(
        session jf: JellyfinSession,
        item: JellyfinItem,
        to directory: URL
    ) async throws -> URL {
        guard let remote = downloadURL(session: jf, itemId: item.Id) else {
            throw JellyfinError.invalidURL
        }
        var req = URLRequest(url: remote)
        req.setValue(authHeader(deviceId: jf.deviceId, token: jf.accessToken), forHTTPHeaderField: "X-Emby-Authorization")
        let (temp, response) = try await session.download(for: req)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw JellyfinError.downloadFailed
        }
        let ext = item.Container?.lowercased() ?? "mp3"
        let safe = item.title
            .replacingOccurrences(of: "/", with: "-")
            .replacingOccurrences(of: ":", with: "-")
        let dest = directory.appendingPathComponent("\(item.Id)_\(safe).\(ext)")
        if FileManager.default.fileExists(atPath: dest.path) {
            try FileManager.default.removeItem(at: dest)
        }
        try FileManager.default.moveItem(at: temp, to: dest)
        return dest
    }

    private func getJSON<T: Decodable>(session jf: JellyfinSession, urlString: String) async throws -> T {
        guard let url = URL(string: urlString) else { throw JellyfinError.invalidURL }
        var req = URLRequest(url: url)
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.setValue(authHeader(deviceId: jf.deviceId, token: jf.accessToken), forHTTPHeaderField: "X-Emby-Authorization")
        let (data, response) = try await session.data(for: req)
        if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            let msg = String(data: data, encoding: .utf8)?.prefix(120) ?? ""
            throw JellyfinError.http(http.statusCode, String(msg))
        }
        do {
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            throw JellyfinError.decode
        }
    }
}
